### PR TITLE
[None][feat] Add KV cache manager V2 support for DSA

### PR DESCRIPTION
**Pull request by yizhang-nv**
## Dev Engineer Review

- Added DSA KV cache manager V2 support while preserving V1 selection.
- Registered indexer K cache storage as a native per-layer `Role.INDEX_KEY` buffer.
- Added V2 page-index conversion for gather, scatter, and paged-logits paths.
- Added FP8 and FP4 indexer storage accounting for cache sizing, quotas, and warm-up estimates.
- Added stable representative-layer selection for coalesced KV cache pools.
- Added draft-cache buffer handling for FlashMLA and DSA CUDA-graph replay.
- Added model defaults that enable KV cache manager V2 for DeepSeek V3.
- Updated configuration validation and documentation for unsupported V2 combinations.
- Updated copyright and license metadata in affected files.
- No test-list files were changed.
- Review should confirm V2 page scaling, FP4 byte accounting, draft-buffer restoration, and V1 fallback behavior across all supported context-parallel configurations.

## QA Engineer Review

Test code changes are present.

Added or updated coverage includes:

- DSA V2 manager selection, page mapping, layer masks, cache sizing, and metadata behavior.
- DSA draft replay and cached full-KV mapping restoration.
- KV cache V2 stable-layer page-table mapping.
- DSA with STAR context parallelism validation.
- Nested model-default preservation.
- DSA accuracy coverage through `TestGLM52.test_nvfp4`.
- C++ custom-op documentation coverage for `indexer_k_cache_scatter_op`.

No corresponding `tests/integration/test_lists/` changes were reported for these tests. Test-list coverage cannot be confirmed from the available changes.

**Verdict: needs follow-up.**

## Description

Port the DSA KV cache manager V2 work from https://github.com/liji-nv/TensorRT-LLM/commit/492422bd512d479d224f0bb04acf18e9ac24c318 onto the latest `main`.

This change:
- Adds `DSACacheManagerV2` selection through `use_kv_cache_manager_v2` while preserving the V1 path.
- Registers the indexer K cache as a native per-layer `Role.INDEX_KEY` buffer managed by V2.
- Makes indexer gather/scatter and paged-logits paths aware of V2 page-index scaling for coalesced pools.
- Accounts for FP8/FP4 indexer storage in `Role.ALL`, runtime quota calculation, and static KV cache warm-up estimation.
- Adds DSA V2 accuracy coverage and targeted unit tests for cache layout and scaled page indices.

## Test Coverage

- Clean C++ source builds passed for SM80 (A100), SM90 (H100/H200), SM100 (B200), and SM103 (B300).
- DSA V2 manager selection, FP8/FP4 layout, role sizing, and warm-up quota: 5 passed.
- Indexer gather/scatter page-index scaling with FP8/FP4: 4 passed.
- DSA V1 scatter regression: 1 passed.
- YAPF, Ruff lint/format, and `git diff --check` passed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by yufeiwu-nv — APPROVED** [review]

File: tensorrt_llm/_torch/speculative/interface.py [hunk 1]
```diff
@@ -159,39 +159,34 @@ def prepare_attn_metadata_for_draft_replay(attn_metadata,
...
-        # Safety clamp: sanitize stale padding entries beyond num_seqs
-        # that may contain negative or out-of-range values, matching the
-        # regular DSA prepare() flow.
-        m.indexer_k_cache_block_offsets.clamp_(min=0)
+        # Keep the target's pinned host buffers untouched while asynchronous
+        # H2D copies from the draft mappings are in flight.  Restoring their
+        # contents in place can race those copies because CPU writes are not
+        # ordered with the CUDA stream.  The fresh host buffers stay attached
+        # to the metadata until after the draft forward has been enqueued.
+        m.host_indexer_k_cache_block_offsets = torch.empty_like(
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/speculative/interface.py]: This allocates three fresh pinned host buffers on every draft replay, i.e., every decode step. cudaHostAlloc is expensive and can implicitly synchronize, and PyTorch's caching host allocator doesn't guarantee a hit. Since the shapes are fixed, consider pre-allocating a dedicated set of draft-side replacement buffers once (double-buffering) at metadata init, and just swap references here.

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 2]
```diff
@@ -376,11 +380,8 @@ def _get_model_kv_cache_manager_cls(
                 )
         return cls
 
-    def _fallback_if_unsupported_kv_cache_manager_v2(
-            self,
-            kv_cache_manager_cls,
-            model_config: ModelConfig,
-            kv_cache_config: Optional[KvCacheConfig] = None):
+    def _fallback_if_unsupported_kv_cache_manager_v2(self, kv_cache_manager_cls,
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/_util.py]: consider renaming to` _validate_or_fallback_kv_cache_manager_v2 `(or similar).

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 3]
```diff
@@ -391,25 +392,26 @@ def _fallback_if_unsupported_kv_cache_manager_v2(
                 incompat.append("kv_connector_manager")
             if self._max_beam_width is not None and self._max_beam_width > 1:
                 incompat.append("beam_width > 1")
+            sparse_attn_config = model_config.sparse_attention_config
+            if (sparse_attn_config is not None
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/_util.py:624]: This PR introduces two user-visible behavior changes: (a) DSA + STAR context parallelism now raises instead of running; (b) sparse-attention models no longer silently fall back from V2 to V1.  Please update doc under `docs/source/features/feature-combination-matrix.md `or `docs/source/features/kvcache.md`

**Review by zhaoyangwang-nvidia — APPROVED** [review]
Approve for **trt-llm-runtime-devs** with few comments.

**Review by yingguo-trt — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/kv_cache_manager_v2.py [hunk 4]
```diff
@@ -1192,6 +1200,15 @@ def _get_block_scale_role(self, role_a: DataRole) -> Optional[DataRole]:
             return None
         return Role.KEY_BLOCK_SCALE
 
+    def _get_pool_layer_id(self, pool_id: int, role: DataRole) -> LayerId:
```
Comment (VALLIS-NERIA) [inline tensorrt_llm/_torch/pyexecutor/kv_cache_manager_v2.py]: This function is not much meaningful

File: tensorrt_llm/_torch/pyexecutor/_util.py (hunk 2 above)
Comment (yizhang-nv) [reply tensorrt_llm/_torch/pyexecutor/_util.py]: Renamed to _validate_or_fallback_kv_cache_manager_v2; both target and draft manager call sites now use that name (4f86a5e75f).

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 5]
```diff
@@ -1,3 +1,6 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
```
[the remark this answers is not in the document]
Comment (yizhang-nv) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py:3]: Fixed in 8a891d7b78: removed the executable bit (100755 to 100644). I also added the requested precise get_model_defaults annotations.

File: tensorrt_llm/_torch/speculative/interface.py (hunk 1 above)
Comment (yizhang-nv) [reply tensorrt_llm/_torch/speculative/interface.py]: Addressed. Dedicated draft-side host/device buffers are now allocated once in DSAtrtllmAttentionMetadata.create_buffers_for_indexer(), and draft replay only rebinds and restores references. The per-step torch.empty_like allocations are gone (e21da307a4, retained in 8a891d7b78).

File: tensorrt_llm/_torch/pyexecutor/_util.py (hunk 3 above)
Comment (yizhang-nv) [reply tensorrt_llm/_torch/pyexecutor/_util.py:624]: Update: the documentation change has been removed from this PR so KV-cache documentation can be updated together in a coordinated follow-up. This PR intentionally keeps the implementation changes only (4282bb13c6).

File: tensorrt_llm/_torch/pyexecutor/kv_cache_manager_v2.py (hunk 4 above)
Comment (yizhang-nv) [reply tensorrt_llm/_torch/pyexecutor/kv_cache_manager_v2.py]: Removed the thin wrapper and inlined the lookup at both call sites in 8a891d7b78. The config-order mapping itself is retained because layer_grouping order is not part of the backend contract; the behavior-level reordered-grouping test remains.

File: tests/unittest/llmapi/test_llm_args.py [hunk 6]
```diff
@@ -3985,3 +3985,40 @@ def test_deepseek_resolves_auto_to_python_on_nixl(self) -> None:
...
+    )
+    llm_args = TorchLlmArgs(
+        model="/tmp/dummy_model",
+        kv_cache_config=KvCacheConfig(use_kv_cache_manager_v2=user_setting),
+    )
+
+    ModelLoader.load_config_and_apply_defaults("/tmp/dummy_model", llm_args,
+                                               checkpoint_loader)
+
+    assert llm_args.kv_cache_config.use_kv_cache_manager_v2 is expected
```
[the remark this answers is not in the document]
Comment (yizhang-nv) [reply tests/unittest/llmapi/test_llm_args.py]: This implementation-detail test was removed in 8a891d7b78 following reviewer feedback, so the requested negative case no longer applies.
